### PR TITLE
Add `prefer-array-flat` rule

### LINKED ISSUE #975: Rule proposal: `prefer-array-flat`
Prefer `Array#flat` over legacy techniques to flatten arrays.

This could be auto-fixable.

## Fail

```js
array.flatMap(x => x);

array.reduce((a, b) => a.concat(b), []);

array.reduce((a, b) => [...a, ...b], []);

[].concat(array);

[].concat(...array);

[].concat.apply([], array);

Array.prototype.concat.apply([], array);

_.flatten(array); // Lodash/Underscore
```

## Pass

```js
array.flat();
```

### PR DESCRIPTION
**Pull request by fisker**
Fixes #975

### CODE REVIEW

**Comment by fisker** [comment]
Half code in this rule just esquery selector. :smile:

File: docs/rules/prefer-array-flat.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Prefer `Array#flat()` over legacy techniques to flatten arrays
+
+ES2019 introduced a new method [`Array#flat()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) that flattens arrays, and there's a [`depth` parameter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat#parameters), so you can flatten deep nesting arrays with it.
```
Comment (sindresorhus) [inline docs/rules/prefer-array-flat.md:3]: I don't think the `depth` parameter is relevant. People can read about it when they click the MDN link.

File: index.js [hunk 2]
```diff
@@ -84,6 +84,9 @@ module.exports = {
 				'unicorn/prefer-add-event-listener': 'error',
 				'unicorn/prefer-array-find': 'error',
 				// TODO: Enable this by default when targeting Node.js 12.
+				// TODO[@fisker]: Set to `off` before merge
```
Comment (sindresorhus) [inline index.js:87]: Don't forget this.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by fisker** [comment]
Added a `functions` option, I think this can be useful https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1126/commits/43e53805541b60a7692574c222ec927f946d874d

**Comment by fisker** [comment]
CI failure should not related, I'll look into it.

File: docs/rules/prefer-array-flat.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	"unicorn/prefer-array-flat": [
+		"error",
+		{
+			"functions": [
+				"flatArray",
+				"utils.flat"
+			]
+		}
+	]
+}
```
Comment (sindresorhus) [inline docs/rules/prefer-array-flat.md:78]: I think our config examples should be in JS, not JSON.

**Pull request merged**